File: ch19p663stateRestoration/p476pageController/Pep.m


#import "Pep.h"

@interface Pep ()
@property (nonatomic, weak) IBOutlet UILabel* name;
@property (nonatomic, weak) IBOutlet UIImageView* pic;
@end

@implementation Pep

// given a Pep boy's name, we display his name and picture

// we are the restoration class
// we are being asked to vend a new basic Pep
// we have encoded which Pep boy was being displayed, so we can call init with that info

+ (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)ic
                                                            coder:(NSCoder *)coder {
    NSLog(@"pep %@", ic);
    UIViewController* result = nil;
    NSString* id = [ic lastObject];
    if ([id isEqualToString:@"pep"]) {
        NSString* boy = [coder decodeObjectForKey:@"boy"];
        result = [[Pep alloc] initWithPepBoy:boy nib:nil bundle:nil];
    }
    return result;
}


-(void)encodeRestorableStateWithCoder:(NSCoder *)coder {
    NSLog(@"pep encode");
    [coder encodeObject:self.boy forKey:@"boy"]; // this is our only state-saving move!
    [super encodeRestorableStateWithCoder:coder];
}

-(void)decodeRestorableStateWithCoder:(NSCoder *)coder {
    NSLog(@"pep decode");
    // nothing further to do
    [super decodeRestorableStateWithCoder:coder];
}

- (id) initWithPepBoy: (NSString*) inputboy nib: (NSString*) nib bundle: (NSBundle*) bundle {
    self = [self initWithNibName:nib bundle:bundle];
    if (self) {
        self->_boy = [inputboy copy];
        self.restorationIdentifier = @"pep"; //
        self.restorationClass = [self class]; //
    }
    return self;
}


- (void)viewDidLoad
{
    [super viewDidLoad];
    self.name.text = self.boy;
    self.pic.image = [UIImage imageNamed:
                      [NSString stringWithFormat: @"%@.jpg", [self.boy lowercaseString]]];
}

- (IBAction)doButton:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: iOS13bookExamples/bk2ch06p357StateSaveAndRestoreWithNSUserActivity/ch19p626pageController/Pep.swift


import UIKit

class Pep: UIViewController {
    
    let boy : String
    @IBOutlet var name : UILabel!
    @IBOutlet var pic : UIImageView!
    
    @IBInspectable var boyName : String = "Manny"
    
    init(pepBoy boy:String) {
        self.boy = boy
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        self.boy = self.boyName
        super.init(coder: coder)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.name.text = self.boy
        self.pic.image = UIImage(named:self.boy.lowercased())
    }
    
    override var description : String {
    return self.boy
    }
    


}
